File: android/src/main/java/com/th3rdwave/safeareacontext/SafeAreaProviderManager.java
package com.th3rdwave.safeareacontext;

import androidx.annotation.NonNull;
import androidx.annotation.Nullable;

import com.facebook.react.bridge.ReactContext;
import com.facebook.react.common.MapBuilder;
import com.facebook.react.uimanager.ThemedReactContext;
import com.facebook.react.uimanager.UIManagerHelper;
import com.facebook.react.uimanager.ViewGroupManager;
import com.facebook.react.uimanager.ViewManagerDelegate;
import com.facebook.react.uimanager.annotations.ReactProp;
import com.facebook.react.viewmanagers.RNCSafeAreaProviderManagerDelegate;
import com.facebook.react.viewmanagers.RNCSafeAreaProviderManagerInterface;

import java.util.Map;

public class SafeAreaProviderManager extends ViewGroupManager<SafeAreaProvider> implements RNCSafeAreaProviderManagerInterface<SafeAreaProvider> {
  private final ViewManagerDelegate<SafeAreaProvider> mDelegate;

  public SafeAreaProviderManager() {
    super();

    mDelegate = new RNCSafeAreaProviderManagerDelegate<>(this);
  }

  @Nullable
  @Override
  protected ViewManagerDelegate<SafeAreaProvider> getDelegate() {
    return mDelegate;
  }

  @Override
  @NonNull
  public String getName() {
    return "RNCSafeAreaProvider";
  }

  @Override
  @NonNull
  public SafeAreaProvider createViewInstance(@NonNull ThemedReactContext context) {
    return new SafeAreaProvider(context);
  }

  @Override
  public Map<String, Object> getExportedCustomDirectEventTypeConstants() {
    return MapBuilder.<String, Object>builder()
      .put(InsetsChangeEvent.EVENT_NAME, MapBuilder.of("registrationName", "onInsetsChange"))
      .build();
  }

  private static final SafeAreaProvider.OnInsetsChangeListener ON_INSETS_CHANGE_LISTENER =
    new SafeAreaProvider.OnInsetsChangeListener() {
      @Override
      public void onInsetsChange(SafeAreaProvider view, EdgeInsets insets, Rect frame) {
        ReactContext reactContext = (ReactContext) view.getContext();

        int reactTag = view.getId();
        UIManagerHelper.getEventDispatcherForReactTag(reactContext, reactTag)
          .dispatchEvent(
            new InsetsChangeEvent(
              UIManagerHelper.getSurfaceId(reactContext), reactTag, insets, frame));
      }
    };

  @Override
  protected void addEventEmitters(@NonNull ThemedReactContext reactContext, @NonNull SafeAreaProvider view) {
    super.addEventEmitters(reactContext, view);

    view.setOnInsetsChangeListener(ON_INSETS_CHANGE_LISTENER);
  }
}

File: android/src/main/java/com/th3rdwave/safeareacontext/SafeAreaProviderManager.kt
package com.th3rdwave.safeareacontext

import com.facebook.react.bridge.ReactContext
import com.facebook.react.module.annotations.ReactModule
import com.facebook.react.uimanager.ThemedReactContext
import com.facebook.react.uimanager.UIManagerHelper
import com.facebook.react.viewmanagers.RNCSafeAreaProviderManagerDelegate
import com.facebook.react.viewmanagers.RNCSafeAreaProviderManagerInterface
import com.facebook.react.views.view.ReactViewGroup
import com.facebook.react.views.view.ReactViewManager

@ReactModule(name = SafeAreaProviderManager.REACT_CLASS)
class SafeAreaProviderManager :
    ReactViewManager(), RNCSafeAreaProviderManagerInterface<ReactViewGroup> {
  private val mDelegate = RNCSafeAreaProviderManagerDelegate(this)

  override fun getDelegate() = mDelegate

  override fun getName() = REACT_CLASS

  override fun createViewInstance(context: ThemedReactContext) = SafeAreaProvider(context)

  override fun getExportedCustomDirectEventTypeConstants() =
      mutableMapOf(
          InsetsChangeEvent.EVENT_NAME to mutableMapOf("registrationName" to "onInsetsChange"))

  override fun addEventEmitters(reactContext: ThemedReactContext, view: ReactViewGroup) {
    super.addEventEmitters(reactContext, view)
    (view as? SafeAreaProvider)?.setOnInsetsChangeHandler(::handleOnInsetsChange)
  }

  companion object {
    const val REACT_CLASS = "RNCSafeAreaProvider"
  }
}

private fun handleOnInsetsChange(view: SafeAreaProvider, insets: EdgeInsets, frame: Rect) {
  val reactContext = view.context as ReactContext
  val reactTag = view.id
  UIManagerHelper.getEventDispatcherForReactTag(reactContext, reactTag)
      ?.dispatchEvent(InsetsChangeEvent(getSurfaceId(reactContext), reactTag, insets, frame))
}
